### DISCUSSION TITLE
how to declare instances for typeclasses with functional dependencies?

### DISCUSSION THREAD

**Discussion by dark-ether** [haskell-effectful/effectful#119] [Q&A] [answered]
i was trying to write a Megaparsec effect following the guide on Effectful.Dispatch.Dynamic and when it became time to write the typeclass it complained that e and s weren't uniquely determined by es which makes sense because it would need to look at the instance context to know that it is. 
i do know i could just use a ParsecT over Eff es but i wanted the possibility of having arbitrary ordering of effects. 

the error i got is the following 
```
Effectful/Megaparsec.hs:47:10: error:
    • Illegal instance declaration for ‘MonadParsec e s (Eff es)’
        The liberal coverage condition fails in class ‘MonadParsec’
          for functional dependency: ‘m -> e s’
        Reason: lhs type ‘Eff es’ does not determine rhs types ‘e’, ‘s’
        Un-determined variables: e, s
    • In the instance declaration for ‘MonadParsec e s (Eff es)’
   |
47 | instance forall e s es. (Ord e ,P.Stream s, Megaparsec e s :> es)=> MonadParsec e s (Eff es ) where
   |          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
```

basically how can i force it to accept my instance declaration given that i know it is safe?

**Comment by arybczak** [comment, marked as answer]
Oh. Indeed, I forgot about this particular case.

You can force GHC to ignore the coverage condition by using [a trick](https://github.com/arybczak/ghc-proposals/blob/covered-pragma/proposals/0000-covered-pragma.rst#motivation), i.e. including the instance head in the context:

```haskell
instance forall e s es. (Ord e ,P.Stream s, Megaparsec e s :> es, MonadParsec e s (Eff es))=> MonadParsec e s (Eff es ) where
```

For the record, I didn't think too much about potential consequences for this particular case, but it should work as expected.

Also, you'll probably want to use effectful-plugin for good user experience with this effect.
